types/pg/index.d.ts

```diff
@@ -105,12 +107,12 @@ export class Pool extends events.EventEmitter {
     end(callback: () => void): void;
 
     query(queryStream: QueryConfig & stream.Readable): stream.Readable;
-    query(queryConfig: QueryArrayConfig): Promise<QueryArrayResult>;
-    query(queryConfig: QueryConfig): Promise<QueryResult>;
-    query(queryText: string, values?: any[]): Promise<QueryResult>;
-    query(queryConfig: QueryArrayConfig, callback: (err: Error, result: QueryArrayResult) => void): Query;
-    query(queryTextOrConfig: string | QueryConfig, callback: (err: Error, result: QueryResult) => void): Query;
-    query(queryText: string, values: any[], callback: (err: Error, result: QueryResult) => void): Query;
+    query<T=any>(queryConfig: QueryArrayConfig): Promise<QueryArrayResult<T>>;
```

Comment:
I'm not sure how comfortable I am with this change - type parameters that aren't observed in the parameter list almost always indicate a code smell because they're really just type assertions.

Reply:
Ah you're right, it's almost a type assertion.
In an other hand, it seems more convenient to use a type parameter than to write a type assertion : 
```ts
// with a type parameter, I just have to pass it
client.query<MyRow>('SELECT NOW()').then(res => {
  const someRows: MyRow[] = res.rows;
});

// with type assertion : 
// - I have to import QueryResult
// - I have to put parenthesis around the argument
// - it's more verbose
import { QueryResult } from 'pg'
client.query('SELECT NOW()').then((res: QueryResult<MyRow>) => {
  const someRows: MyRow[] = res.rows;
});

I have to admit I don't totally get why using a type parameter for a type assertion would hurt.

Anyway, tell me if you want me to remove concerning the promise api, and only keep the changes concerning the callback api.

Otherwise, I pushed the change for pg-ears allowing ts 2.3

Reply:
With the type assertion I would have thought it would just be:

```ts
client.query('SELECT NOW()').then(res => {
  const someRows: MyRow[] = res.rows;
});
```

Why is the ceremony with `QueryResult` required?

Reply:
@bradleyayers if you do this, res.rows isn't typed.

But perhaps indeed we could write : 
```ts
client.query('SELECT NOW()').then(res => {
  const someRows = res.rows as MyRow[]
});
```
but I still think it's more intuitive to pass a type parameter to a function (because I expect a typed result "returned" by this function) rather than to

Reply:
I find using type parameters in this way to be quite convenient, but I'm really curious about @DanielRosenwasser's hint that it's a code smell. Where can I learn more about this?

Reply:
> I find using type parameters in this way to be quite convenient, but I'm really curious about @DanielRosenwasser's hint that it's a code smell. Where can I learn more about this?

Same here, just curious to know if I missed something.